File: BOJ/Stack/BJ9012.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin')
  .toString()
  .trim()
  .split('\n');

const T = +input.shift();

for(const br of input) {
  const stack = [];
  for(let i=0; i<br.length; i++) {
    if(br[i] === '(' ) stack.push('(');
    else {
      if(stack.length > 0) stack.pop();
      else {
        stack.push('(');
        break;
      }  
    }
  } 
  console.log(stack.length === 0 ? 'YES' : 'NO');
}

File: BOJ/Stack/BJ9012.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int T;

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  
  cin >> T;
  string str;
  for(int i=0; i<T; i++) {
    cin >> str;

    int size = (int) str.size();
    stack<char> s;
    for(int j=0; j<size; j++) {
      if(str[j]=='(') s.push('(');
      else {
        if(s.size() > 0) s.pop();
        else {
          s.push('(');
          break;
        }
      }
    } 

    string ans;
    if(s.size() == 0) ans = "YES";
    else ans = "NO";
    cout << ans << '\n';
  }
  return 0;
}